### PR TITLE
cephfs_mirror: increment sync_failures when sync_perms() and sync_snaps() fails

### PR DESCRIPTION
**Pull request by joscollin**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/tools/cephfs_mirror/PeerReplayer.cc [hunk 1]
```diff
@@ -1588,6 +1588,9 @@ void PeerReplayer::run(SnapshotReplayerThread *replayer) {
 	  r = sync_perms(*dir_root);
 	  if (r < 0) {
 	    _inc_failed_count(*dir_root);
+	    if (m_perf_counters) {
```
Comment (vshankar) [inline src/tools/cephfs_mirror/PeerReplayer.cc:1591]: Suggest we do this in a single place in this source here thereby removing the one ins sync_snaps() and handling the failure cases in run().

**Review by vshankar — CHANGES_REQUESTED** [review]
Good catch!

File: src/tools/cephfs_mirror/PeerReplayer.cc (hunk 1 above)
Comment (joscollin) [reply src/tools/cephfs_mirror/PeerReplayer.cc]: @vshankar Fixed. Please check.

File: src/tools/cephfs_mirror/PeerReplayer.cc [hunk 2]
```diff
@@ -1585,13 +1566,26 @@ void PeerReplayer::run(SnapshotReplayerThread *replayer) {
...
         int r = register_directory(*dir_root, replayer);
         if (r == 0) {
-	  r = sync_perms(*dir_root);
-	  if (r < 0) {
-	    _inc_failed_count(*dir_root);
-	  } else {
-	    sync_snaps(*dir_root, locker);
-	  }
-	  unregister_directory(*dir_root);
+          r = sync_perms(*dir_root);
```
[the remark this answers is not in the document]
Comment (joscollin) [reply src/tools/cephfs_mirror/PeerReplayer.cc:1569]: So you want to keep the sync_snaps() function, return `r` and just have the `m_perf_counters->inc(l_cephfs_mirror_peer_replayer_snap_sync_failures);` in the `run` function?

[the remark this answers is not in the document]
Comment (joscollin) [reply src/tools/cephfs_mirror/PeerReplayer.cc:1586]: updated

File: src/tools/cephfs_mirror/PeerReplayer.cc [hunk 3]
```diff
@@ -1585,13 +1583,19 @@ void PeerReplayer::run(SnapshotReplayerThread *replayer) {
...
+          if (r < 0) {
```
[the remark this answers is not in the document]
Comment (joscollin) [reply src/tools/cephfs_mirror/PeerReplayer.cc]: Fixed

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
@rishabh-d-dave Please do qa

**Comment by vshankar** [comment]
@joscollin please create a tracker and link it in the commit message. This needs to be backported.

**Review by vshankar — APPROVED** [review]

**Comment by joscollin** [comment]
> @joscollin please create a tracker and link it in the commit message. This needs to be backported.

@vshankar Linked https://tracker.ceph.com/issues/65345.

**Comment by joscollin** [comment]
jenkins test api

File: src/tools/cephfs_mirror/PeerReplayer.cc [hunk 4]
```diff
@@ -1758,13 +1756,20 @@ void PeerReplayer::run(SnapshotReplayerThread *replayer) {
...
-	  }
-	  unregister_directory(*dir_root);
+          r = sync_perms(*dir_root);
+          if (r == 0) {
+            r = sync_snaps(*dir_root, locker);
+            if (r < 0 && m_perf_counters) {
+              m_perf_counters->inc(l_cephfs_mirror_peer_replayer_snap_sync_failures);
+            }
+          }
+          else if (r < 0) {
```
Comment (vshankar) [inline src/tools/cephfs_mirror/PeerReplayer.cc:1766]: does this needs to be `else if`. Seems like just `else` would suffice.

Comment (joscollin) [reply src/tools/cephfs_mirror/PeerReplayer.cc]: Pushed the change. `sync_perms` returns  `<= 0` only.

**Comment by joscollin** [comment]
jenkins test make check

File: src/tools/cephfs_mirror/PeerReplayer.cc [hunk 5]
```diff
@@ -1758,13 +1756,20 @@ void PeerReplayer::run(SnapshotReplayerThread *replayer) {
...
-	  }
-	  unregister_directory(*dir_root);
+          r = sync_perms(*dir_root);
+          if (r == 0) {
+            r = sync_snaps(*dir_root, locker);
+            if (r < 0 && m_perf_counters) {
+              m_perf_counters->inc(l_cephfs_mirror_peer_replayer_snap_sync_failures);
+            }
+          }
+          else {
```
Comment (vshankar) [inline src/tools/cephfs_mirror/PeerReplayer.cc:1766]: nit: } else {

**Review by vshankar — APPROVED** [review]

**Comment by joscollin** [comment]
jenkins test make check

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#2024-05-13

**Pull request merged**